crates/starfish/core/src/authority_service.rs

```diff
@@ -522,31 +579,52 @@ impl<C: CoreThreadDispatcher> NetworkService for AuthorityService<C> {
             .with_label_values(&[peer_hostname])
             .inc();
 
-        // 7. Add additional headers from bundle to dag, receive missing ancestors for
+        // 7. Add digests to filter. Exclude from the vector those that are already
+        //    inserted
+        let mut digests_to_add_to_filter = vec![];
+        for block_header in additional_block_headers.iter() {
+            digests_to_add_to_filter.push(block_header.digest())
+        }
+        digests_to_add_to_filter.push(verified_block.digest());
+        let digests_to_exclude = self
+            .received_block_headers
+            .add_batch(digests_to_add_to_filter)
+            .await;
+        let mut index = 0;
+        additional_block_headers.retain(|block_header| {
```

Comment:
Does this work though? I guess this only works if it's guaranteed that the relative ordering in both vectors is the same. I think it would be safer to implement it like:

```
additional_block_headers.retain(|block_header| {
    !digests_to_exclude.iter().any(|digest| block_header.digest() == *digest)
});
```